src/main/java/com/vmware/vipclient/i18n/messages/service/PatternService.java

```diff
@@ -4,89 +4,107 @@
  */
 package com.vmware.vipclient.i18n.messages.service;
 
-import java.util.Locale;
-
+import com.vmware.vipclient.i18n.VIPCfg;
+import com.vmware.vipclient.i18n.base.DataSourceEnum;
+import com.vmware.vipclient.i18n.util.LocaleUtility;
 import org.json.simple.JSONObject;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import com.vmware.vipclient.i18n.VIPCfg;
-import com.vmware.vipclient.i18n.base.DataSourceEnum;
-import com.vmware.vipclient.i18n.messages.api.opt.local.LocalPatternOpt;
-import com.vmware.vipclient.i18n.messages.api.opt.server.RemotePatternOpt;
-import com.vmware.vipclient.i18n.util.ConstantsKeys;
-import com.vmware.vipclient.i18n.util.LocaleUtility;
+import java.util.ListIterator;
+import java.util.Locale;
 
 /**
  * The class represents date formatting
  */
 public class PatternService {
     Logger logger = LoggerFactory.getLogger(PatternService.class);
+    private static final String PATTERNS_PREFIX = "patterns_";
 
     public JSONObject getPatternsByCategory(String locale, String category) {
         JSONObject patterns = getPatterns(locale);
         return (JSONObject) patterns.get(category);
     }
 
     public JSONObject getPatterns(String locale) {
+        if(locale == null || locale.isEmpty()) {
+            logger.warn("Locale is empty!");
+            return null;
+        }
+        locale = locale.replace("_", "-").toLowerCase();
         JSONObject patterns = null;
-        logger.debug("Look for pattern from cache!");
-        patterns = new PatternCacheService().lookForPatternsFromCache(locale);// key
-        if (patterns == null) {
-            patterns = getPatternsFromBundle(locale);
-            if ((patterns == null) && !LocaleUtility.isDefaultLocale(locale)) {
-                patterns = getPatternsFromBundle(ConstantsKeys.EN);
-            }
-            if (null != patterns) {
-                logger.info("Got the pattern  with   locale [{}].\n", locale);// [datetime] and
-                logger.info("Cache pattern!\n\n");
-                new PatternCacheService().addPatterns(locale, patterns);
-            }
+        logger.debug("Look for pattern from cache for locale [{}]!", locale);
+        String cacheKey = PATTERNS_PREFIX + locale;
+        patterns = new PatternCacheService().lookForPatternsFromCache(cacheKey);// key
+        if (patterns != null) {
+            logger.debug("Find pattern from cache for locale [{}]!", locale);
+            return patterns;
+        }
+        patterns = getPatternsFromDS(locale, VIPCfg.getInstance().getMsgOriginsQueue().listIterator());
+        if (patterns != null) {
+            logger.debug("Find the pattern for locale [{}].\n", locale);// [datetime] and
+            new PatternCacheService().addPatterns(cacheKey, patterns);
+            logger.debug("Pattern is cached for locale [{}]!\n\n", locale);
+            return patterns;
+        }
+        if (!LocaleUtility.isDefaultLocale(locale)) {
```

Comment:
Put this fallback logic in PatternMessage so that this method here in PatternService is always for the requested language only.

Once you move the fallback logic into PatternMessage, update the cache in there in such a way that the key with the requested language is mapped to the FormatCacheItem of the fallback language.
`new PatternCacheService().addPatterns(<keyOfRequestedLanguage>, <patternsOfDefaultLanguage>);`

Without the above proposed changes, your code will always try to fetch the requested locale first before using the fallback locale that is already in the cache. With the above proposed changes, any succeeding calls for the requested language will immediately return the cached fallback locale until it expires.

Note: You can either put the above logic in PatternMessage as I propose (just like how it is in TranslationMessage.getMessages), or create another method here in PatternService to keep the logic in the "service layer".